### PR TITLE
Provide the ability to store more information about listeners in the emitter object

### PR DESCRIPTION
**Pull request by djerius**
In my application emitters require more information about listeners than just the callback coderef.
The background for this is that I'm implementing one-to-one sending of events in addition to the existing one-to-many distribution, and the emitter needs to know which callback to invoke for a given subscriber.  

This changeset provides a general interface to attaching extra data for each listener in the cache.
I've introduced a `Beam::Listener` class which `subscribe()` inserts into the `listener` attribute.  I've copied the interface provided by `emit()` to allow the caller to use a child class of `Beam::Listener` as well as extra attributes.

In order that the emitter have access to this, I've renamed `_listener` to `listener` and thus made it part of the public API.

I've not yet written tests, so this PR is not yet ready, but I'd like to know if this approach is something you'd incorporate.

### CODE REVIEW

**Comment by preaction** [comment]
To start out, yes, I think the core concept of a Listener object is best placed here inside the core Beam::Emitter. I'd like to see, instead of the `_listener` data structure being directly exposed, a `listener()` method that takes a single argument (the event to list) and returns a list of listeners in order (exposing data structures locks us in to keeping those exact data structures).

Some background: Beam::Emitter is similar in concept to the event emitter base class in Adobe's ActionScript 3. That's why Events are objects, and have methods like `stop` and `prevent_default`. The Beam framework as a whole is an attempt to bring Java-like integration classes into the Perl space (Beam::Wire is inversion of control with static config files similar to Swing). There are works in progress like Beam::Runner (https://github.com/preaction/Beam-Runner) to provide distributed processing built on top of Beam::Emitter and Beam::Wire.

From that angle, having a class to encapsulate a waiting listener is the correct thing. It enables introspection. Beam::Listener objects could record a stack trace when they are created for debugging event handler exceptions (this alone is, for me, reason enough to do this). There are all kinds of fun things one could do with the abstraction layer, including making event listeners into a distributed worker farm (like you might be doing). 

But Beam::Emitter is pretty specifically a one-to-many fan-out. Changing the `emit()` method to support one-to-one based on an `isa()` check of the listener's class is a deep change to make that will involve practically copy/paste of the original `emit()` method. So, I worry that though Beam::Listener is a good idea, it might not get you to where you want to go.

TL;DR: I'm for this, but how is it going to help your use-case?

**Comment by djerius** [comment]
Here is some [working code](https://github.com/preaction/Beam-Emitter/files/590435/nrp.txt) using the patched version of `Beam::Emitter`. 

The base idea is that there is a "network" of peers who attach (subscribe) and detach (unsubscribe) themselves from each other.  Each node contains data which may be a function of its peers' data.  When a node is detached, its subscribers are notified so that they can adjust their data.  A node may detach from all of its peers (which can be done without the patch), or from individual peers.  It's the latter which requires the additional metadata for listeners and their introspection.

The design/interface is a bit inverted from `Beam::Emitter`, as nodes have an attach method which subscribes them to other nodes' events, and a detach method to unsubscribe themselves.E.g.,

`  $n1->attach( $n2, $name )`
and
`  $n1->detach( $n2, $name )`

Callbacks are by default methods on `$n1`.  There's some added complexity to ensure that callback loops are detected.

Rather than extend the `emit()` method (which in my mind implies broadcast) I've added a `send()` method, which is given the peer to which to send the event. 

Your suggestion of providing a method to return listeners by event name is a much better interface. I'll implement that.

**Comment by preaction** [comment]
Okay, so the integration point is the Listener and Event classes, the current Emitter role is used for listener management (because that stays the same), and some new methods were added to your class to send messages to the right subscribers. So we're good to go ahead here.

**Comment by djerius** [comment]
Concise summary; sorry to have thrown the whole code at you.

The latest commits add the `listeners()` method, revert exposing `_listeners`, add tests and documentation.

`listeners()` returns a ref to the actual array.  I didn't document that, as I don't whether you view the ability for the caller to manipulate it as a bug or a feature.

File: t/listeners.t [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+# test constructor is being called with args
+throws_ok(
+    sub { $foo->subscribe( evt2 => $s22, class => 'Goo' ) },
+    qr/missing required arguments/i,
+    "missing attr for custom listener class"
+);
+
+my $us22 = $foo->subscribe( evt2 => $s22, class => 'Goo', attr => 's22' );
+
+{
```
Comment (preaction) [inline t/listeners.t]: For each of these naked blocks in the test, could you modify them slightly to use `subtest` and provide a short description of what the test is intended to demonstrate? Something like: `subtest "Ensure the correct listener object is removed from the list" => sub { ... };`. All the Beam stuff relies on a version of Test::More that has `subtest`.

File: lib/Beam/Emitter.pm [hunk 2]
```diff
@@ -15,20 +15,27 @@ use Beam::Listener;
...
+specified event from this emitter.  The array elements are instances
+of sub-classes of L<Beam::Listener>.
+
+=cut
+
+sub listeners {
+
+    my ( $self, $name ) = @_;
+
+    return $self->_listeners->{$name} || [];
```
Comment (preaction) [inline lib/Beam/Emitter.pm]: Could you deref this arrayref and return the list instead? Returning the arrayref from `listeners()` could be a potential source of spooky bugs: It could cause unrelated code to think it owned that ref and could do what it wanted (`shift`, `splice`, etc...). If that ref gets passed around too far, the idea of who owns that ref gets lost. It's similar to C/C++ pointer ownership: The location of the error message (exception in Perl, segfault in C/C++) gets so far removed from the actual source of the problem that it becomes extremely difficult to debug. It's not so much that it's more or less useful to return a list or an arrayref, but it's protection against a rare corner case that leads to frustration.

It'd probably be easiest to deref the array and just return the list of listeners. Then people can make it into an array or an arrayref if they feel. In general I prefer returning lists to refs (both array and hash) mostly for that reason: Ownership is hard to know and even harder to fix.

If the emitter itself wants to edit that ref, it has the private `_listeners` hash to do so (though we should probably discourage that so that the data structure isn't part of the API). And if someone else wants to remove a listener, they could pass the listener into `unsubscribe`. In fact, the changes in this PR could enable us in the future to make that even easier to do...

**Comment by preaction** [comment]
Also, sorry about the conflict(s). Trying to clear out the open tickets so I can bundle this all up in a single release and blog about it.

File: lib/Beam/Emitter.pm (hunk 2 above)
Comment (djerius) [inline lib/Beam/Emitter.pm]: makes sense. will do.

Comment (djerius) [inline lib/Beam/Emitter.pm]: Another school of thought is to return arrayrefs rather than lists in order to reduce stack size when there's the possibility of returning a large number of items.  The returned arrayref is a new one, e.g. [ @{$arrayref} ].

File: t/listeners.t (hunk 1 above)
Comment (djerius) [inline t/listeners.t]: done

**Comment by djerius** [comment]
rebased, so should be conflict free.

**Comment by preaction** [comment]
Thanks! The travis failures look like they're my fault, so I'll fix them and then I can merge this. Do you have a website / blog / company allowing you to work on this / consultancy website / something I should link to in my blog post about these changes?

**Comment by djerius** [comment]
Well, I work for the Smithsonian Astrophysical Observatory at the [Chandra X-Ray Center](http://cxc.harvard.edu/) in the [Optics Calibration group](http://cxc.harvard.edu/cal/Hrma/Index.html).  I intend to use this code to fix some design problems in [PDLx::Mask](https://metacpan.org/pod/PDLx::Mask) which I will use to fix design issues in [PDLx::Algorithm::Center](https://github.com/djerius/PDLx-Algorithm-Center) which I'm using in internal code to help understand the contribution of [this](http://cxc.cfa.harvard.edu/contrib/juda/memos/hrc_pileup/index.html) to [this](http://cxc.harvard.edu/cal/Hrc/PSF/empPSF.html).  That's a mouthful.

If any of that is helpful, please use it.  I'm one of the few people around here still using Perl for numerical analysis (99% are using Python and NumPy).

**Comment by preaction** [comment]
This is merged to master, and I'll likely tag a release today or tomorrow after I address #9. There might even be some kind of interesting interaction with #9: The Beam::Listener API could allow for the subref returned by `callback` to have error-handling built-in to it, so that the listener could decide how it wants to handle errors in its callback, and then the emitter has the stuff in #9 to decide how it wants to handle errors in its listeners' callbacks...

And yes, absolutely, that's great! I love PDL and anything I can do to show people using it is a good thing ;) I used to do quantitative time series analysis at a bank with PDL and TA-lib, but they also moved to Python, NumPy, and Pandas (which, to be fair, Pandas is cool).

File: lib/Beam/Emitter.pm (hunk 2 above)
Comment (preaction) [inline lib/Beam/Emitter.pm]: Yeah, that's a possible problem, but I don't think usage of this will reach that point. We're likely talking many thousands of items before that becomes a true bottleneck (though now I am curious to run some microbenchmarks to see...). Besides, making a shallow copy like that does the same thing, really: all the things in `$arrayref` go on the stack as it's dereferenced in list context, then the stack gets used to create the arrayref. I imagine `perl` makes some optimizations in both situations, and until I need otherwise, I generally trust `perl` to do it as fast as it thinks it can. So it's a tradeoff between potential bugs or potential performance problems, and at the moment I think those potential bugs are a bit more potential ;)

Were this something like an ETL or something that we could very obviously pass off ownership of the arrayref and had good reason to believe that the data set would be extremely large, I'd likely fall the other way. Or, if Perl had built-in and fast immutable iterable like Python's tuple...
